File: Advanced/snakesandladders-cordapp/contracts/src/main/java/net/corda/samples/snl/contracts/GameBoardContract.java
package net.corda.sample.snl.contracts;

import net.corda.core.contracts.CommandData;
import net.corda.core.contracts.Contract;
import net.corda.core.transactions.LedgerTransaction;
import net.corda.sample.snl.states.BoardConfig;
import net.corda.sample.snl.states.GameBoard;
import org.jetbrains.annotations.NotNull;

public class GameBoardContract implements Contract {

    @Override
    public void verify(@NotNull LedgerTransaction tx) throws IllegalArgumentException {
        // Contract verification logic should be implemented here
        if(tx.getCommands().size() != 1)
            throw new IllegalArgumentException("One command Expected");

        if(tx.getCommand(0).getValue() instanceof Commands.Create){
            verifyCreate(tx);
        }else if(tx.getCommand(0).getValue() instanceof Commands.PlayMove){
            verifyPlay(tx);
        }

    }

    private void verifyCreate(LedgerTransaction tx) throws IllegalArgumentException{
        // Contract verification logic for create game should be implemented here
        if(tx.getOutputStates().size() != 1 || tx.getInputStates().size() != 0)
            throw new IllegalArgumentException("Zero Input and One Output Expected");

        if(!(tx.getOutput(0) instanceof GameBoard))
            throw new IllegalArgumentException("Output of type GameBoard expected");


    }

    private void  verifyPlay(LedgerTransaction tx) throws IllegalArgumentException {
        // Contract verification logic for play move should be implemented here

        if(tx.getReferences().size() == 0 || !(tx.getReferenceInput(0) instanceof BoardConfig)){
            throw new IllegalArgumentException("One reference Input of BoardConfig Expected");
        }

        if(tx.getOutputStates().size() != 1 || tx.getInputStates().size() != 1)
            throw new IllegalArgumentException("One Input and One Output Expected");

    }

    public interface Commands extends CommandData {
        class Create implements Commands {}
        class PlayMove implements Commands {
            private int roll;

            public PlayMove(int roll) {
                this.roll = roll;
            }

            public int getRoll() {
                return roll;
            }
        }
    }
}

File: kotlin-samples/Advanced/snakesandladders-cordapp/contracts/src/main/kotlin/net/corda/samples/snl/contracts/GameBoardContract.kt
package net.corda.samples.snl.contracts


import net.corda.core.contracts.CommandData
import net.corda.core.contracts.Contract
import net.corda.core.transactions.LedgerTransaction
import net.corda.samples.snl.states.BoardConfig
import net.corda.samples.snl.states.GameBoard


class GameBoardContract : Contract {

    companion object {
        var ID = "net.corda.samples.snl.contracts.GameBoardContract"
    }

    @Throws(IllegalArgumentException::class)
    override fun verify(tx: LedgerTransaction) {
        // Contract verification logic should be implemented here
        if (tx.commands.size != 1) throw IllegalArgumentException("One command Expected")
        if (tx.getCommand<CommandData>(0).value is Commands.Create) {
            verifyCreate(tx)
        } else if (tx.getCommand<CommandData>(0).value is Commands.PlayMove) {
            verifyPlay(tx)
        }
    }

    @Throws(IllegalArgumentException::class)
    private fun verifyCreate(tx: LedgerTransaction) {
        // Contract verification logic for create game should be implemented here
        if (tx.outputStates.size != 1 || tx.inputStates.size != 0) throw IllegalArgumentException("Zero Input and One Output Expected")
        if (tx.getOutput(0) !is GameBoard) throw IllegalArgumentException("Output of type GameBoard expected")
    }

    @Throws(IllegalArgumentException::class)
    private fun verifyPlay(tx: LedgerTransaction) {
        // Contract verification logic for play move should be implemented here
        if (tx.references.size == 0 || tx.getReferenceInput(0) !is BoardConfig) {
            throw IllegalArgumentException("One reference Input of BoardConfig Expected")
        }
        if (tx.outputStates.size != 1 || tx.inputStates.size != 1) throw IllegalArgumentException("One Input and One Output Expected")
    }

    interface Commands : CommandData {
        class Create : Commands
        class PlayMove(val roll: Int) : Commands
    }
}
